executor/cook/io_helper.py

```diff
@@ -73,49 +78,80 @@ def print_err(string_data, flush=False, newline=True):
     -------
     Nothing.
     """
+    stderr_buffer = sys.stderr.buffer
     with __stderr_lock__:
-        sys.stderr.write(string_data)
+        if isinstance(data, str):
+            stderr_buffer.write(data.encode())
+        else:
+            stderr_buffer.write(data)
         if newline:
-            sys.stderr.write(os.linesep)
+            stderr_buffer.write(os.linesep)
         if flush:
-            sys.stderr.flush()
+            stderr_buffer.flush()
 
 
-def process_output(label, out_file, max_bytes_read_per_line, out_fn, flush_fn):
+def process_output(label, process, out_fd, out_fn, flush_fn, min_reads_before_flush, max_read_nb_chunk_size):
     """Processes output piped from the out_file and prints it using the out_fn function.
     When done reading the file, calls flush_fn to flush the output.
 
     Parameters
     ----------
     label: string
         The string to associate in outputs
-    out_file: a file object
+    out_fd: a file descriptor
         Provides that output from the child process
-    max_bytes_read_per_line: int
-        The maximum number of bytes to read per call to readline().
     out_fn: function(string)
         Function to output a string
     flush_fn: function()
         Function that flushes the output.
+    min_reads_before_flush: int
+        The number of read attempts before triggering a flush
+    max_read_nb_chunk_size: int
+        The maximum number of bytes to read per call to readline().
 
     Returns
     -------
     Nothing.
     """
     logging.info('Starting to pipe {}'.format(label))
+
+    def safe_flush():
+        try:
+            flush_fn()
+        except:
+            logging.exception('Error while flushing contents {}'.format(label))
+
+    def nb_read(read_pipe, timeout_secs=0.10):
+        read_ready, _, _ = select.select([read_pipe], [], [], timeout_secs)
+        if read_ready:
+            return os.read(read_pipe, max_read_nb_chunk_size)
+        return None
+
+    reads_since_flush = 0
+
+    bytes_in_buffer = 0
     try:
         while True:
-            line = out_file.readline(max_bytes_read_per_line)
-            if not line:
+            data = nb_read(out_fd)
+            reads_since_flush += 1
+            if data is not None:
+                logging.info('Read: {} [label={}]'.format(data, label))
```

Comment:
Do we want to log every read at info?

Reply:
Reverted the commit, so not relevant anymore.